### ISSUE TITLE
Handling dates for seasons

### ISSUE THREAD

**Issue by F3PiX** [rails-girls-summer-of-code/rgsoc-teams#301]
RubyNuby here, so proceed with caution! 
I think it works like this:
Orga's can add a Season in orga/seasons/new, including starting and closing dates for applications, the coding summer, etc
But in the code, the dates are hardcoded, AND will always override the input from the new-form (AFAIK).
 
I was thinking: what if the Season dates will be set only in the orga/seasons/new form? We can add default dates in the form fields. The orga's won't have to type more letters, because of the defaults. But, if the dates change (as was the case in 2015 submissions), the dates are very easy to update in the form. 
Super nice selfish bonus: I can easily switch seasons so that I can work on the Application Form and on other stuff. 

What I do not oversee: if the cure is worse than the problem. 

(Related to the updated Readme: Quickstart Guide for beginners - setting the season open/close)

**Comment by sareg0** [comment]
Hey F3PiX,

This is a very interesting feature/improvement I'd love to talk further about. In theory, I think it sounds great! 

I'm trying to think of any edge cases to account for, as well as more support for why this could be a good idea (in addition to your points above).

Thoughts from other stakeholders welcome :)

**Comment by carpodaster** [comment]
Hey @F3PiX, can you give me an example where the dates are hardcoded? If they are, _that_ feels like a bug. Do you refer to [app/models/season.rb:28](https://github.com/rails-girls-summer-of-code/rgsoc-teams/blob/master/app/models/season.rb#L28)? The default values are only set if we didn't pass something else (ie. via the seasons form). The default values are in place to ensure we always have a working season with a meaningful dates without the need to edit them first (e.g. after a fresh checkout of the app).

That being said, a developer-mode to quickly switch between seasons would be super lecker of course!

**Comment by F3PiX** [comment]
O, it is totally possible that I am confusing two different things. 
Yes, I found the default values, but the way I (!!) handled it, my new_seasons data were overridden. Not in the lines you linked, I know, because I stubbed them. And maybe not hardcoded, I start to doubt that now.  But they were overridden. 

What happened: I created a new season 2016 in the orga/seasons form, with applications dates in Aug 2015.  That makes that the link to Apply Now appears in the menu. That's cool. But when I click the link, I got an error message : Applications 2015 are closed. (I am still figuring this out, and I think it has nothing to do with this issue, but with my teams season.)

I issued the issue at this point in my search, and found that it would be cool if 'we developers' could easily switch seasons - easily the key word here :-) 

I see the need of a working season. That is also only for development, isn't it? Could that be solved with adding a season seed?

**Comment by F3PiX** [comment]
No hardcoding to be found! Sorry, my mistake. 
What I found is that there are quite a few different places where 'something happens' to the seasons. It is a bit hard to follow, but I'll do my best. I need to ask the coaches if I should solve this first, before I can add the switch. I'll keep you posted.
(And when that is solved: shouldn't we build a state machine for the season's stages? :-) )

**Comment by carpodaster** [comment]
> What happened: I created a new season 2016 in the orga/seasons form, with applications dates in Aug 2015. That makes that the link to Apply Now appears in the menu. That's cool. But when I click the link, I got an error message : Applications 2015 are closed.

The _current_ season is always looked up in the year, assuming we won't run two SoCs per year. Creating a 2016 season will still return the 2015 season for `Season.current` (unless you modify your Rails machine's time to next year).

Warping oneself into a different phase / state of the season requires modifying the various datetime values for the _current_ season.

Changing the season would require an easy way to pick the phase you as a developer would like to see yourself in and set all datetime values appropriately.

> That is also only for development, isn't it? Could that be solved with adding a season seed?

I'm against putting that into the seeds.rb. The application will still run w/o users, job offers, and what have you, but it will outright crash without a season. Auto-creating a season is simple enough so I'd like to keep it that way.

**Comment by F3PiX** [comment]
@carpodaster 
Totally understand now that seeds is a bad option. Thanks for the clear explanation why. WIL: never use seeding to replace code. (Note to self: "DUH!")

I am still working on following the different season settings through the code.

**Comment by carpodaster** [comment]
Hey @F3PiX, your season switch PR solves this problem, I think. It does not warp into a future season, but it achieves the desired effect by modifying the _current_ one. Do you think this one can be closed?

**Comment by F3PiX** [comment]
Sure! May I have the honor?
